lang/src/org/partiql/lang/planner/transforms/VariableIdAllocator.kt

```diff
@@ -0,0 +1,45 @@
+package org.partiql.lang.planner.transforms
+
+import org.partiql.lang.domains.PartiqlLogical
+import org.partiql.lang.domains.PartiqlLogicalResolved
+
+/**
+ * Allocates register indexes for all local variables in the plan.
+ *
+ * Returns pair containing a logical plan where all `var_decl`s have a [VARIABLE_ID_META_TAG] meta indicating the
+ * variable index (which can be utilized later when establishing variable scoping) and list of all local variables
+ * declared within the plan, which becomes the `locals` sub-node of the `plan` node.
+ */
+internal fun PartiqlLogical.Plan.allocateVariableIds(): Pair<PartiqlLogical.Plan, List<PartiqlLogicalResolved.LocalVariable>> {
+
+    var allLocals = mutableListOf<PartiqlLogicalResolved.LocalVariable>()
+    val planWithAllocatedVariables = VariableIdAllocator(allLocals).transformPlan(this)
+    return planWithAllocatedVariables to allLocals.toList()
+}
+
+private const val VARIABLE_ID_META_TAG = "\$variable_id"
```

Comment:
Nit.: I'm thinking if we need to get all the constants in one place under `planner`.

Reply:
In general, I'd rather not have one file with a bunch of unrelated constants.  Note that this is `private`, putting it somewhere else means it can't be `private` anymore and will pollute the package namespace.  Putting the constant here also has the advantage that the definition is close to its uses, which improves readability IMHO.